lib/client/db/postgres/repl/repl_test.go

```diff
@@ -0,0 +1,418 @@
+	}
+	return acc.String()
+}
+
+func readLine(t *testing.T, c *testCtx) string {
+	t.Helper()
+
+	var n int
+	buf := make([]byte, 1024)
+	// Given that the test connections are piped a problem with the writer side
+	// would lead into blocking reading. To avoid this scenario we're using
+	// the Eventually just to ensure a timeout on reading from the connections.
+	require.EventuallyWithT(t, func(collect *assert.CollectT) {
+		var err error
+		n, err = c.conn.Read(buf[0:])
+		require.NoError(t, err)
+		require.Greater(t, n, 0)
+	}, 5*time.Second, time.Millisecond)
+	return string(buf[:n])
+}
+
+type testCtx struct {
+	cfg        *testCtxConfig
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+
+	// conn is the connection used by tests to read/write from/to the REPL.
+	conn net.Conn
+	// clientConn is the connection passed to the REPL.
+	clientConn net.Conn
+	// serverConn is the fake database server connection (that works as a
+	// PostgreSQL instance).
+	serverConn net.Conn
+
+	route    clientproto.RouteToDatabase
+	pgClient *pgproto3.Backend
+	errChan  chan error
+	// queryChan handling custom queries is enabled the queries received by the
+	// test server will be sent to this channel.
+	queryChan chan string
+}
+
+type testCtxConfig struct {
+	// skipREPLRun when set to true the REPL instance won't be executed.
+	skipREPLRun bool
+	// handleCustomQueries when set to true the PostgreSQL test server will
+	// accept any query sent and reply with success.
+	handleCustomQueries bool
+}
+
+// testCtxOption represents a testCtx option.
+type testCtxOption func(*testCtxConfig)
+
+// WithCustomQueries enables sending custom queries to the PostgreSQL test
+// server. Note that when it is enabled, callers must consume the queries on the
+// query channel.
+func WithCustomQueries() testCtxOption {
+	return func(cfg *testCtxConfig) {
+		cfg.handleCustomQueries = true
+	}
+}
+
+// WithSkipREPLRun disables automatically running the REPL instance.
+func WithSkipREPLRun() testCtxOption {
+	return func(cfg *testCtxConfig) {
+		cfg.skipREPLRun = true
+	}
+}
+
+// StartWithServer starts a REPL instance with a PostgreSQL test server capable
+// of receiving and replying to queries.
+func StartWithServer(t *testing.T, ctx context.Context, opts ...testCtxOption) (*REPL, *testCtx) {
+	t.Helper()
+
+	cfg := &testCtxConfig{}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	conn, clientConn := net.Pipe()
+	serverConn, pgConn := net.Pipe()
+	client := pgproto3.NewBackend(pgproto3.NewChunkReader(pgConn), pgConn)
+	ctx, cancelFunc := context.WithCancel(ctx)
+	tc := &testCtx{
+		cfg:        cfg,
+		ctx:        ctx,
+		cancelFunc: cancelFunc,
+		conn:       conn,
+		clientConn: clientConn,
+		serverConn: serverConn,
+		pgClient:   client,
+		errChan:    make(chan error, 1),
+		queryChan:  make(chan string),
+	}
+
+	t.Cleanup(func() {
+		tc.close()
+
+		select {
+		case err := <-tc.errChan:
+			require.NoError(t, err)
+		case <-time.After(time.Second):
```

Comment:
`time.Second` is probably going to be flaky in CI. I'd increase to 3 or 5 seconds.